File: Data Structures & Algorithms/count-good-nodes-in-binary-tree/submission-2.py
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right

class Solution:
    def goodNodes(self, root: TreeNode) -> int:
        ans = []
        if not root: return ans
        stack = [(root, root.val)]
        while len(stack):
            ele = stack.pop()
            node = ele[0]
            max_value = ele[1]

            if max_value <= node.val:
                ans.append(node.val)
            if node.left: 
                stack.append((node.left, max(max_value, node.left.val)))
            if node.right:
                stack.append((node.right, max(max_value, node.right.val)))
            
        return len(ans)

File: Data Structures & Algorithms/daily-temperatures/submission-2.cpp
class Solution {
public:
    vector<int> dailyTemperatures(vector<int>& temperatures) {
        int n = temperatures.size();
        stack<tuple<int, int>> st;
        vector<int> ans(n, 0);

        st.push(make_tuple(temperatures[n-1], n-1));

        for (int i = n -2; i >= 0; i--)
        {
            while (!st.empty() && temperatures[i] >= get<0>(st.top()))
            {
                st.pop();
            }

            if (st.empty())
            {
                ans[i] = 0;
            }
            else 
            {
                ans[i] = get<1>(st.top()) - i;
            }
            st.push(make_tuple(temperatures[i], i));
        }
        return ans;
    }
};
